src/borrower/feed/navfeed.sol

```diff
@@ -272,70 +282,69 @@ contract NAVFeed is BaseNFTFeed, Interest, Buckets, FixedPoint {
             fv = calcFutureValue(loan, debt, maturityDate_, recoveryRatePD[risk[nftID_]].value);
-            buckets[maturityDate_].value = safeAdd(buckets[maturityDate_].value, fv);
+            buckets[maturityDate_] = safeAdd(buckets[maturityDate_], fv);
         }
 
         futureValue[nftID_] = fv;
 
-        // remove buckets if no remaining assets
-        if (buckets[maturityDate_].value == 0 && firstBucket != 0) {
-            removeBucket(maturityDate_);
-        }
-
         // return decrease NAV amount
-        if (block.timestamp < maturityDate_) {
-            return calcDiscount(safeSub(preFutureValue, fv), uniqueDayTimestamp(block.timestamp), maturityDate_);
-        }
+        return calcDiscount(safeSub(preFutureValue, fv), uniqueDayTimestamp(block.timestamp), maturityDate_);
 
-        // if a loan is overdue the portfolio value is equal to the existing debt multiplied with a write off factor
-        return amount;
     }
 
-    function calcDiscount(uint amount, uint normalizedBlockTimestamp, uint maturityDate_) public view returns (uint result) {
-        return rdiv(amount, rpow(discountRate.value, safeSub(maturityDate_, normalizedBlockTimestamp), ONE));
+    function calcDiscount(uint fv, uint normalizedBlockTimestamp, uint maturityDate_) public view returns (uint result) {
+        return rdiv(fv, rpow(discountRate.value, safeSub(maturityDate_, normalizedBlockTimestamp), ONE));
     }
 
+    function secureSub(uint x, uint y) public pure returns(uint) {
+        if(y > x) {
+            return 0;
+        }
+        return safeSub(x, y);
+    }
 
-    /// calculates the total discount of all buckets with a timestamp > block.timestamp
-    function calcTotalDiscount() public view returns(uint) {
-        uint normalizedBlockTimestamp = uniqueDayTimestamp(block.timestamp);
-        uint sum = 0;
-
-        uint currDate = normalizedBlockTimestamp;
-
-        if (currDate > lastBucket) {
+    function currentDiscount() public view returns(uint) {
+        if (latestDiscount == 0) {
             return 0;
         }
 
-        // only buckets after the block.timestamp are relevant for the discount
-        // assuming its more gas efficient to iterate over time to find the first one instead of iterating the list from the beginning
-        // not true if buckets are only sparsely populated over long periods of time
-        while(buckets[currDate].next == 0) { currDate = currDate + 1 days; }
+        uint nnow = uniqueDayTimestamp(block.timestamp);
+        uint nLastUpdate = uniqueDayTimestamp(lastNAVUpdate);
+
+        uint totalDiscount = rmul(latestDiscount, rpow(discountRate.value, safeSub(nnow, nLastUpdate), ONE));
 
-        while(currDate != NullDate)
-        {
-            sum = safeAdd(sum, calcDiscount(buckets[currDate].value, normalizedBlockTimestamp, currDate));
-            currDate = buckets[currDate].next;
+        uint diff = 0;
+        for(uint i = nLastUpdate; i < nnow; i = i + 1 days) {
```

Comment:
I think this loop could use a comment explaining its intention, it took me a while to figure out. If I understand it correctly:

```suggestion
        // Loop over the loans which matured in between the last NAV update and now.
        // Then remove their discounted future value from the total discount as they are overdue.
        for(uint i = nLastUpdate; i < nnow; i = i + 1 days) {
```

Reply:
Yes,  you are correct. Technically we iterate over the buckets which are loans grouped by different maturity dates

Reply:
We can get rid of the the loop & remove the overdue loans i a sep writeoff function